### PR TITLE
More UTC

### PR DESCRIPTION
**Pull request by gcampo88**
- updates the timestamps in schema for all models to reflect utc datetime nature
- adds a pre-validation Timex helper to ensure that all date field params in a Campaign are utc-able (ie, have hours and minutes defined on the datetime) before they're cast into a Campaign changeset
- removes docker-compose installation from buildspec

### CODE REVIEW

**Comment by gcampo88** [comment]
per slack conversation with @cavis who looked into this: 

- [x] need to update `start_date` and `end_date` column types to `utc_datetime` in a migration -- they weren't actually included in https://github.com/PRX/jingle.prx.org/pull/73 (oops)
- [x] change Date / DateTime structs back to strings, and let Ecto do the parsing (`%{day: 17, month: 4, year: 2010}` to  `"2010-04-17"`)
- [x] on datetime strings you’ll probably want to use iso8601 format `2010-04-17T00:00:00Z`

File: test/controllers/api/campaign_controller_test.exs [hunk 1]
```diff
@@ -2,7 +2,15 @@ defmodule Jingle.API.CampaignControllerTest do
   use Jingle.ConnCase
 
   alias Jingle.Campaign
-  @valid_attrs %{original_copy: "some content", end_date: %{day: 17, month: 4, year: 2010}, podcast_id: 42, sponsor_id: 42, start_date: %{day: 17, month: 4, year: 2010}, zone: "some content"}
+  @valid_attrs %{
+    original_copy: "some content",
+    zone: "some content",
+    podcast_id: 42,
+    sponsor_id: 42,
+    end_date: %{day: 17, month: 4, year: 2010, hour: 0, minute: 0},
```
Comment (cavis) [inline test/controllers/api/campaign_controller_test.exs]: Shouldn't this be a string?  (Or at least use the official struct so it gives you missing-key errors ... `%DateTime{day: 17 ...}`)

File: priv/repo/seeds.exs [hunk 2]
```diff
@@ -44,29 +44,31 @@ dogtalk = Repo.insert! %Podcast{
...
-zip_campaign = Repo.insert! %Campaign{
+zip_campaign = Campaign.changeset(%Campaign{}, %{
   sponsor_id: zip_recruiter.id,
   podcast_id: dogtalk.id,
-  start_date: Ecto.Date.cast!("2017-06-27"),
-  end_date: Ecto.Date.cast!("2017-07-20"),
-  due_date: Ecto.Date.cast!("2017-07-20"),
+  start_date: %{day: 17, month: 4, year: 2010},
+  end_date: %{day: 17, month: 4, year: 2010},
+  due_date: %{day: 17, month: 4, year: 2010},
```
Comment (cavis) [inline priv/repo/seeds.exs]: I'm sort of surprised this works, without minute/hour on there.  But wouldn't it be cleaner to just use iso8601 strings here?

File: web/models/campaign.ex [hunk 3]
```diff
@@ -15,17 +15,36 @@ defmodule Jingle.Campaign do
...
-    |> cast(params, [:sponsor_id, :podcast_id, :start_date, :end_date, :due_date, :original_copy, :edited_copy, :must_say, :notes, :approved, :zone])
-    |> validate_required([:start_date, :end_date, :original_copy, :zone])
+    |> cast(utc_compatible_dates(params), [:sponsor_id, :podcast_id, :start_date, :end_date, :due_date, :original_copy, :edited_copy, :must_say, :notes, :approved, :zone])
+    |> validate_required([:start_date, :end_date, :due_date, :original_copy, :zone])
     |> assoc_constraint(:sponsor)
     |> assoc_constraint(:podcast)
   end
+
+  defp utc_compatible_dates(params) do
+    date_format = "{YYYY}-{0M}-{0D}"
```
Comment (cavis) [inline web/models/campaign.ex]: So it seems like this is either going to (1) parse a plain date-string to a `%NaiveDateTime`, or (2) rescue and pass the string through to Ecto's built in parser.

Is the advantage to (2) just to let the Ecto changeset handle passing any "invalid date" errors on through the controller?

File: web/models/campaign.ex [hunk 4]
```diff
@@ -15,17 +15,36 @@ defmodule Jingle.Campaign do
...
+    try do
+      params = atomize_map(params)
+      %{ params | start_date: Timex.parse!(params[:start_date], date_format),
+                  end_date: Timex.parse!(params[:end_date], date_format),
+                  due_date: Timex.parse!(params[:due_date], date_format) }
+
+    rescue
```
Comment (cavis) [inline web/models/campaign.ex]: It's actually more idiomatic in Elixir to not throw errors, but instead match the `:ok` / `:error` in the tuple.  Something like:

```
case Timex.parse(date_str, date_format) do
  {:ok, dtim} -> dtim,
  {:error, _} -> date_str
end
```

And you may also want to convert those naive datetimes ... there's an example of something similar in [the Castle plugs](https://github.com/PRX/castle.prx.org/blob/master/lib/plugs/interval/time_to.ex).

File: priv/repo/seeds.exs (hunk 2 above)
Comment (cavis) [reply priv/repo/seeds.exs]: And actually, this file doesn't run for me.  (And it runs even less if I change them to strings).

File: web/models/campaign.ex [hunk 5]
```diff
@@ -15,17 +15,36 @@ defmodule Jingle.Campaign do
...
+      # if we errored out on parsing, yield params as they are
+      FunctionClauseError -> params
+    end
+  end
+
+  defp atomize_map(map) do
```
Comment (cavis) [inline web/models/campaign.ex]: nice!

**Review by cavis — CHANGES_REQUESTED** [review]
Phew - I had to read a lot about Ecto to get through this!  Lot of tricky things you figured out in there.  I think I've got 2 general comments on this:

1. Worth discussing whether your interface for the model changeset uses structs or strings.  (What kind of date should you be passing in to the changeset?)
2. Generally don't want to throw/catch errors in elixir.

File: priv/repo/seeds.exs (hunk 2 above)
Comment (gcampo88) [reply priv/repo/seeds.exs]: 😶  oops

File: web/models/campaign.ex (hunk 3 above)
Comment (gcampo88) [reply web/models/campaign.ex]: made some changes, but re: 2) -- yes

File: web/models/campaign.ex (hunk 4 above)
Comment (gcampo88) [reply web/models/campaign.ex]: :+1:

**Comment by gcampo88** [comment]
ok @cavis I thiiiink I addressed all the things...

File: test/models/campaign_test.exs [hunk 6]
```diff
@@ -3,14 +3,52 @@ defmodule Jingle.CampaignTest do
...
+    map_date_params = %{
+      start_date: %{day: 17, month: 4, year: 2017, hour: 0, minute: 0},
+      end_date: %{day: 17, month: 4, year: 2019, hour: 0, minute: 0},
+      due_date: %{day: 10, month: 4, year: 2017, hour: 0, minute: 0}
+    }
+    changeset = Campaign.changeset(%Campaign{}, Map.merge(required, map_date_params))
+    assert changeset.valid?
+    assert changeset.changes.end_date.year == 2019
+    #
+    tuple_date_params = %{
```
Comment (cavis) [inline test/models/campaign_test.exs]: What's the difference between this test and the one before it?

File: web/models/campaign.ex [hunk 7]
```diff
@@ -15,17 +15,36 @@ defmodule Jingle.Campaign do
...
+    |> cast(utc_compatible_dates(params), [:sponsor_id, :podcast_id, :start_date, :end_date, :due_date, :original_copy, :edited_copy, :must_say, :notes, :approved, :zone])
+    |> validate_required([:start_date, :end_date, :due_date, :original_copy, :zone])
     |> assoc_constraint(:sponsor)
     |> assoc_constraint(:podcast)
   end
+
+  defp utc_compatible_dates(params) do
+    params = atomize_map(params)
+    date_params = Map.take(params, [:start_date, :end_date, :due_date])
+    parsed_date_params = for { k, v } <- date_params, into: %{}, do: { k, parse_dtim(v) }
```
Comment (cavis) [inline web/models/campaign.ex:35]: Nice.  Seems like this handles the case where not all of the date-keys are even in the params.

**Review by cavis — APPROVED** [review]
One question about a test... but looks good to me.  :+1000:

File: web/models/campaign.ex (hunk 7 above)
Comment (gcampo88) [reply web/models/campaign.ex:35]: exactly -- if that happens, we want the validation to catch it, not an error here

File: test/models/campaign_test.exs (hunk 6 above)
Comment (gcampo88) [reply test/models/campaign_test.exs]: oops. nothing! fixing.

**Pull request merged**